### PR TITLE
Fix/multiple monitors with same identifiers

### PR DESCRIPTION
**Pull request by BingoWon**
This commit addresses the issue of multiple monitors with exactly the same identifiers, acknowledging that they can exist for valid reasons under certain circumstances. It introduces an option for users to decide whether to filter out these duplicate monitors.

In my specific scenario on Windows 11, I have two monitors with identical identifiers, including EDID, serial number, and name. The latest version of the software is unable to handle this situation appropriately as it always filters out the second monitor with the same identifier, resulting in improper functioning of all operations on the second monitor.

However, duplicate monitors can be differentiated by their system-assigned indices, which makes this commit possible.

Please note that the tests included in this repository have only been tested and passed on Windows 11 so far.

Why are there monitors sharing exactly the same identifiers?
In short, small manufacturers don't care.
They tend not to place a high value on branding and aim to compete primarily on price. Their focus is typically on assembly, and they absolutely do not possess the core technology. They survive and thrive. Their monitors work fine as well.

### CODE REVIEW

File: screen_brightness_control/__init__.py [hunk 1]
```diff
@@ -233,7 +240,7 @@ def fade_brightness(
 
 
 def list_monitors_info(
-    method: Optional[str] = None, allow_duplicates: bool = False, unsupported: bool = False
+    method: Optional[str] = None, allow_duplicates: bool = ALLOW_DUPLICATES, unsupported: bool = False
```
Comment (Crozzers) [inline screen_brightness_control/__init__.py:243]: Not sure what this change is for since changing `ALLOW_DUPLICATES` won't change this parameters default value after the library is imported.
```python
>>> FLAG = False
>>> def a(param=FLAG):
...     print(param)
... 
>>> a()
False
>>> FLAG = True
>>> a()
False
```

File: screen_brightness_control/__init__.py [hunk 2]
```diff
@@ -17,6 +17,13 @@
 _logger = logging.getLogger(__name__)
 _logger.addHandler(logging.NullHandler())
 
+# Global variable to control whether duplicate monitors are allowed
```
Comment (Crozzers) [inline screen_brightness_control/__init__.py:20]: If this is refactored as a docstring under the variable, our documentation generator can pick it up for the online docs

File: screen_brightness_control/__init__.py [hunk 3]
```diff
@@ -17,6 +17,13 @@
...
+# Duplicate monitor info could occur due to two known reasons
+# 1. The same monitor is connected through different interfaces (e.g., HDMI, DisplayPort, DVI, VGA) at the same time.
+# 2. Different monitors share the exactly same identifiers, including EDID, serial number and name, due to careless manufacturer.
+# In case #1, duplicate monitor info is considered redundant and should be filtered out.
+# In case #2, it is essential and should be preserved.
+ALLOW_DUPLICATES = False
```
Comment (Crozzers) [inline screen_brightness_control/__init__.py:26]: Not sure if having this via kwargs and global variables is ideal, especially since changing this variable won't change what the kwarg defaults to.
```python
>>> import inspect
>>> inspect.signature(sbc.list_monitors_info)
<Signature (method: Optional[str] = None, allow_duplicates: bool = False, unsupported: bool = False) -> List[dict]>
>>> sbc.ALLOW_DUPLICATES = True
>>> inspect.signature(sbc.list_monitors_info)
<Signature (method: Optional[str] = None, allow_duplicates: bool = False, unsupported: bool = False) -> List[dict]>
```
I think this would get pretty confusing for users and that maybe instead an `allow_duplicates` kwarg should be added to relevant functions and propagated by them, although that is a pretty long list:
- `get_brightness`
- `set_brightness`
- `fade_brightness`
  - Will be absorbed into `**kwargs` automatically but needs to be passed on to `filter_monitors`
- `filter_monitors`
- `__brightness`
- `list_monitors`

**Review by Crozzers — COMMENTED** [review]
Hi there, apologies for the delayed reply and thank you for taking the time to address this issue.
I've left a couple of feedback/question comments mostly around the use of global vars but the PR concept seems solid to me.

**Review by Crozzers — COMMENTED** [review]
> This approach removes the need to pass 'allow_duplicates=allow_duplicates' int about 7 locations within init.py

But to avoid this you've had to add a decorator, decorate 6 different functions and add an `allow_duplicates` parameter to those 6 functions anyway.

I feel like this PR is trying to solve two different problems.
1. The duplicates problem
2. The tedium of propagating and repeatedly passing parameters around

This library is mostly stateless (since #3), with all of the current "state" residing in the current call stack. Trying to add thread-safe global state to this is going to be a challenge.

> improving code cleanliness by reducing parameter passing in about 8 locations

I do understand this problem, it's caused issues before (eg: ffa10a1) and we do it with other parameters too (eg: `display`, `method`, `force`). It's tedious and very verbose but I think it's something that warrants its own separate discussion on how to address it.

**Comment by BingoWon** [comment]
@Crozzers Thank you for your feedback.

The latest commit addresses the thread safety concern highlighted in the previous review. Although using a thread lock would technically solve the issue, it would reintroduce blocking into threads that were initially non-blocking. The cause of this side effect might not be immediately apparent to users and they would not have any options to prevent this, making this solution less than ideal.

The 'allow_duplicates' parameter has now been added to all functions that could directly or indirectly call the 'filter_monitor' function. This straightforward and effective approach has been accepted by all reviewers participating in this PR discussion.

The global variable 'ALLOW_DUPLICATES' has been retained. This variable does not alter existing usage patterns, thereby ensuring backward compatibility. It simply provides users with a new, global option.

A decorator has been implemented to ensure the proper functioning of the 'ALLOW_DUPLICATES' global variable. This decorator has been applied to all user-facing functions, prioritizing the 'allow_duplicates' parameter to offer ample flexibility.

Please note that all tests included in this repository have been executed and passed exclusively on Windows 11 as of now.

**Comment by Crozzers** [comment]
I'm happy to merge this but I don't think the global variable and decorator fit well as top-level components in the library. Furthermore, if all of this infrastructure is going to be added for global variables I'd like to see it used more than once.

My proposal is that the global variable functionality is moved to a `config` submodule and that support is added for other widely applicable parameters (eg: `method`, `force`, etc...). The usage would then look something like this:
```python
import screen_brightness_control as sbc
sbc.config.ALLOW_DUPLICATES = True
sbc.config.FORCE = False
[...]
```

This is somewhat out of scope for this PR so I'm happy to merge this and refactor it afterwards if that's an agreeable plan.

**Comment by Crozzers** [comment]
Merged. Thanks for this.

I'll try to play around with the config submodule this weekend and I'll let you know how that goes
